### PR TITLE
resource/aws_elasticache_cluster: Introduce initial CustomizeDiff

### LINKED ISSUE #2421: Downgrading Redis engine version should force new resource
### Terraform Version
Terraform v0.10.8
AWS provider v1.3.0

### Affected Resource(s)
`aws_elasticache_cluster`

### Terraform Configuration Files
```hcl
resource "aws_elasticache_cluster" "instance" {
  engine = "redis"
  engine_version = "3.2.10"
  ...
```

```hcl
resource "aws_elasticache_cluster" "instance" {
  engine = "redis"
  engine_version = "3.2.4"
  ...
```
### Expected Behavior
Plan should show the resource needs to be destroyed and recreated, not updated in place.

### Actual Behavior
Terraform plan shows update in place planned.
```
~ module.redis.aws_elasticache_cluster.instance
      engine_version:     "3.2.10" => "3.2.4"
```
and apply fails because AWS doesn't seem to allow updating a running node like that.
```
aws_elasticache_cluster.instance: [WARN] Error updating ElastiCache cluster (***), error: InvalidParameterCombination: Cannot modify redis from 3.2.10 to 3.2.4
```
> Important, You can upgrade to newer engine versions (see Upgrading Engine Versions), but you cannot dwngrade to older engine versions except by deleting the existing cluster and creating it anew.

http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/Clusters.Modify.html

### PR DESCRIPTION
**Pull request by bflad**
Closes #1702
Closes #2421

* Support plan time validation of az_mode
* Support plan time validation of node_type requiring VPC for cache.t2 instances
* Support plan time validation of num_cache_nodes > 1 for redis
* ForceNew memcached on node_type changes
* ForceNew on engine_version downgrades

### CODE REVIEW

File: aws/resource_aws_elasticache_cluster.go [hunk 1]
```diff
@@ -210,6 +217,81 @@ func resourceAwsElasticacheCluster() *schema.Resource {
 		},
 
 		Schema: resourceSchema,
+
+		CustomizeDiff: customdiff.Sequence(
+			func(diff *schema.ResourceDiff, v interface{}) error {
+				// Plan time validation for az_mode
+				// InvalidParameterCombination: Must specify at least two cache nodes in order to specify AZ Mode of 'cross-az'.
+				azMode, azModeOk := diff.GetOk("az_mode")
```
Comment (paultyng) [inline aws/resource_aws_elasticache_cluster.go]: Not something to hold up merging, and probably leaning more towards preference, but you could rewrite this a little so you could just reuse the standard `ok` and short circuiting a little earlier (also helping the line length a little):

```golang
if azMode, ok := diff.GetOk("az_mode"); !ok || azMode.(string) != elasticache.AZModeCrossAz {
  return nil
}

if numCacheNodes, ok := diff.GetOk("num_cache_nodes"); !ok || numCacheNodes.(int) != 1 {
  return nil
}

return errors.New(...)
```

File: aws/resource_aws_elasticache_cluster.go [hunk 2]
```diff
@@ -210,6 +217,81 @@ func resourceAwsElasticacheCluster() *schema.Resource {
...
+				if azModeOk && numCacheNodesOk && azMode.(string) == elasticache.AZModeCrossAz && numCacheNodes.(int) == 1 {
+					return errors.New(`az_mode "cross-az" is not supported with num_cache_nodes = 1`)
+				}
+				return nil
+			},
+			func(diff *schema.ResourceDiff, v interface{}) error {
+				// Plan time validation for engine_version
+				// InvalidParameterCombination: Cannot modify memcached from 1.4.33 to 1.4.24
+				// InvalidParameterCombination: Cannot modify redis from 3.2.6 to 3.2.4
+				if diff.Id() != "" && diff.HasChange("engine_version") {
```
Comment (paultyng) [inline aws/resource_aws_elasticache_cluster.go]: Similar to the other short circuiting one, this flattens out if you just do the easy return first:

```golang
if diff.Id() == "" || !diff.HasChange("engine_version") {
  return nil
}

// rest of it, but flatter indentation
```

**Comment by bflad** [comment]
Short-circuit logic is passing acceptance testing:
```
13 tests passed (all tests)
=== RUN   TestAccAWSElasticacheCluster_NumCacheNodes_Redis_Ec2Classic
--- PASS: TestAccAWSElasticacheCluster_NumCacheNodes_Redis_Ec2Classic (3.06s)
=== RUN   TestAccAWSElasticacheCluster_AZMode_Redis_Ec2Classic
--- PASS: TestAccAWSElasticacheCluster_AZMode_Redis_Ec2Classic (492.09s)
=== RUN   TestAccAWSElasticacheCluster_AZMode_Memcached_Ec2Classic
--- PASS: TestAccAWSElasticacheCluster_AZMode_Memcached_Ec2Classic (504.87s)
=== RUN   TestAccAWSElasticacheCluster_basic
--- PASS: TestAccAWSElasticacheCluster_basic (529.05s)
=== RUN   TestAccAWSElasticacheCluster_importBasic
--- PASS: TestAccAWSElasticacheCluster_importBasic (544.75s)
=== RUN   TestAccAWSElasticacheCluster_multiAZInVpc
--- PASS: TestAccAWSElasticacheCluster_multiAZInVpc (627.27s)
=== RUN   TestAccAWSElasticacheCluster_vpc
--- PASS: TestAccAWSElasticacheCluster_vpc (678.90s)
=== RUN   TestAccAWSElasticacheCluster_NodeTypeResize_Redis_Ec2Classic
--- PASS: TestAccAWSElasticacheCluster_NodeTypeResize_Redis_Ec2Classic (816.04s)
=== RUN   TestAccAWSElasticacheCluster_NodeTypeResize_Memcached_Ec2Classic
--- PASS: TestAccAWSElasticacheCluster_NodeTypeResize_Memcached_Ec2Classic (918.65s)
=== RUN   TestAccAWSElasticacheCluster_EngineVersion_Memcached_Ec2Classic
--- PASS: TestAccAWSElasticacheCluster_EngineVersion_Memcached_Ec2Classic (1071.12s)
=== RUN   TestAccAWSElasticacheCluster_decreasingCacheNodes
--- PASS: TestAccAWSElasticacheCluster_decreasingCacheNodes (1133.37s)
=== RUN   TestAccAWSElasticacheCluster_EngineVersion_Redis_Ec2Classic
--- PASS: TestAccAWSElasticacheCluster_EngineVersion_Redis_Ec2Classic (1145.27s)
=== RUN   TestAccAWSElasticacheCluster_snapshotsWithUpdates
--- PASS: TestAccAWSElasticacheCluster_snapshotsWithUpdates (1352.50s)
```

Merging!

**Comment by bflad** [comment]
This has been released in [version 1.12.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1120-march-23-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**